### PR TITLE
Speedup ContextInstances::removeEach if there are null instances

### PR DESCRIPTION
**Pull request by franz1981**
This change was due to:

image

which has a compute as

image

in short: is the case where there's a generated `ContextInstances` with N > 1 instances fields but just a single one computed (ie not null).
Before this PR we are going to perform 3 * N atomic volatile ops (using strong memory barriers), despite just 1 would be necessary. 
The 3 atomic ops are:
ie:
1. lock
2. volatile store `null` on the instance field
3. unlock

@mkouba @Ladicek 

I have no idea which cutoff value make sense here and, as said in the previous comment at

> 2. volatile store `null` on the instance field

this is another small problem, which could be solved by using `VarHandle::setRelease(..., null)` on the volatile instance field, while holding the lock, but I would raise this in a separate PR.
TBH this one is already decent, because it saves the 3 atomic ops to happen in one go; but is assuming that when a value is `null` is a terminal behavior and if a compute happen concurrently, `remove` can still observe it as uninitialized.

My only concern ATM is that we have (regardless this PR) a lot of "undefined" states in the 2 `ContextInstances` impls, which are prone to both optimizations/concurrency misbehaviour, and I would like, in a separate issue, to state which ones are admissible/wrong/rare but still admissible based on how we use things.

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java [hunk 1]
```diff
@@ -92,11 +93,20 @@ Collection<Resource> generate(DotName scope) {
...
+        // not all instances are being computed, yet.
+        if (idToFields.size() > 16) {
+            implementRemoveEach(contextInstances, idToFields);
 
-        // These methods are needed to significantly reduce the size of the stack map table for getAllPresent() and removeEach()
-        implementLockAll(contextInstances, idToFields);
-        implementUnlockAll(contextInstances, idToFields);
+            // These methods are needed to significantly reduce the size of the stack map table for getAllPresent() and removeEach()
+            implementLockAll(contextInstances, idToFields);
+            implementUnlockAll(contextInstances, idToFields);
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:106]: The `lockAll()` and `unlockAll()` methods are also used in `getAllPresent()`, so either you need to rewrite `getAllPresent()` in a similar fashion, or you need to keep `lockAll()` and `unlockAll()`.

Comment (franz1981) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:106]: thanks for checking, let me fix it, good point!

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java [hunk 2]
```diff
@@ -149,6 +159,42 @@ private void implementUnlockAll(ClassCreator contextInstances, Map<String, Insta
...
+            ResultHandle lock = isNotNull.readInstanceField(fields.lock, isNotNull.getThis());
+            isNotNull.invokeInterfaceMethod(MethodDescriptors.LOCK_LOCK, lock);
+            isNotNull.assign(copy, isNotNull.readInstanceField(fields.instance, isNotNull.getThis()));
+            isNotNull.writeInstanceField(fields.instance, isNotNull.getThis(), isNotNull.loadNull());
+            isNotNull.invokeInterfaceMethod(MethodDescriptors.LOCK_UNLOCK, lock);
+        }
+        BytecodeCreator consumerIsNotNull = removeEach.ifNotNull(removeEach.getMethodParam(0)).trueBranch();
+        for (ResultHandle result : results) {
+            consumerIsNotNull.ifNotNull(result).trueBranch().invokeInterfaceMethod(MethodDescriptors.CONSUMER_ACCEPT,
+                    removeEach.getMethodParam(0), result);
```
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java:193]: This should be

```suggestion
                    consumerIsNotNull.getMethodParam(0), result);
```

but it probably doesn't matter.

**Comment by franz1981** [comment]
@mkouba what surprise me (mostly because I'm very ignorant on the topic) is that an empty quarkus application (not even an hello world endpoint) still generate 2 `ContextInstances`:
- Request scoped with 3 fields
- Application scoped with 6 fields

I would like to understand what is deciding how many fields should be created, if possible

**Comment by mkouba** [comment]
> @mkouba what surprise me (mostly because I'm very ignorant on the topic) is that an empty quarkus application (not even an hello world endpoint) still generate 2 `ContextInstances`:
> 
>     * Request scoped with 3 fields
> 
>     * Application scoped with 6 fields
> 
> 
> I would like to understand what is deciding how many fields should be created, if possible

Each field represents a bean of the given scope. Since your app is empty these beans are provided by the extensions. They are also [unremovable](https://quarkus.io/guides/cdi-reference#remove_unused_beans); either by definition or marked by an extension.

**Comment by franz1981** [comment]
@mkouba @Ladicek no idea if the failures are related or not...

**Comment by franz1981** [comment]
@Ladicek @mkouba putting this on draft: this could be a better and more complete fix -> https://github.com/franz1981/quarkus/commits/main_lazy_ctx_instances/

**Pull request closed without merging**